components/butterfly/src/server.rs

```diff
@@ -628,22 +598,72 @@ impl Server {
+            });
+        self.service_config_store
+            .with_rumors_transient_rsw(&service_group, Transient::refresh);
+        self.service_file_store
+            .with_rumors_transient_rsw(&service_group, Transient::refresh);
+        self.election_store
+            .with_rumors_transient_rsw(&service_group, |r| {
+                if *self.member_id == r.member_id {
+                    r.refresh();
+                }
+            });
+        self.update_store
+            .with_rumors_transient_rsw(&service_group, |r| {
+                if *self.member_id == r.member_id {
+                    r.refresh();
+                }
+            });
+    }
+
+    /// Purge any expired rumors from any of our rumor stores, including Members
+    pub fn purge_expired(&mut self, expiration_date: DateTime<Utc>) {
+        self.election_store.purge_expired_rsr(expiration_date);
+        self.update_store.purge_expired_rsr(expiration_date);
+        self.service_store.purge_expired_rsr(expiration_date);
+        self.service_config_store.purge_expired_rsr(expiration_date);
+        self.service_file_store.purge_expired_rsr(expiration_date);
+        self.member_list.purge_expired_mlw(expiration_date);
+    }
+
+    /// Gather and return RumorKeys for all our rumors
+    pub fn rumor_keys(&self) -> Vec<RumorKey> {
+        let mut keys = self.service_store.lock_rsr().live_keys();
+        keys.append(&mut self.service_config_store.lock_rsr().live_keys());
+        keys.append(&mut self.service_file_store.lock_rsr().live_keys());
+        keys.append(&mut self.election_store.lock_rsr().live_keys());
+        keys.append(&mut self.update_store.lock_rsr().live_keys());
+        keys.append(&mut self.departure_store.keys_rsr());
+        keys.append(&mut self.member_list.rumor_keys_mlr());
+        keys
+    }
+
     /// Given a membership record and some health, insert it into the Member List.
     ///
     /// # Locking (see locking.md)
     /// * `MemberList::entries` (write)
-    fn insert_member_from_rumor_mlw(&self, member: Member, mut health: Health) {
-        let rk: RumorKey = RumorKey::from(&member);
-        if member.id == self.member_id() && health != Health::Alive {
-            let mut me = self.member.write().expect("Member lock is poisoned");
-            if member.incarnation >= me.incarnation() {
-                me.refute_incarnation(member.incarnation);
-                health = Health::Alive;
+    fn insert_member_from_rumor_mlw(&self, mut member: Member, mut health: Health) {
+        if member.id == self.member_id() {
+            if health != Health::Alive {
+                let mut me = self.member.write().expect("Member lock is poisoned");
+                if member.incarnation >= me.incarnation() {
+                    me.refute_incarnation(member.incarnation);
+                    health = Health::Alive;
+                }
+            }
+
+            if member.expired(Utc::now()) {
+                member.refresh();
```

Comment:
I'm not following your reasoning on why it's unexpected.  The `Member` rumor needs to get refreshed somehow, or it will expire, right?  I'm happy to add comments, but I'm wondering if there's a better place to put this that's less unexpected.  This felt like the most appropriate place to me.